src/ui/public/agg_response/hierarchical/build_hierarchical_data.js

```diff
@@ -17,99 +17,53 @@
  * under the License.
  */
 
-import _ from 'lodash';
-import { toastNotifications } from 'ui/notify';
-import { extractBuckets } from './_extract_buckets';
-import { createRawData } from './_create_raw_data';
-import { arrayToLinkedList } from './_array_to_linked_list';
-import AggConfigResult from '../../vis/agg_config_result';
-import { AggResponseHierarchicalBuildSplitProvider } from './_build_split';
+import { toArray } from 'lodash';
 import { HierarchicalTooltipFormatterProvider } from './_hierarchical_tooltip_formatter';
 
 export function BuildHierarchicalDataProvider(Private) {
-  const buildSplit = Private(AggResponseHierarchicalBuildSplitProvider);
   const tooltipFormatter = Private(HierarchicalTooltipFormatterProvider);
 
-  return function (vis, resp) {
-    // Create a reference to the buckets
-    let buckets = vis.getAggConfig().bySchemaGroup.buckets;
-
-    // Find the metric so it's easier to reference.
-    // TODO: Change this to support multiple metrics.
-    const metric = vis.getAggConfig().bySchemaGroup.metrics[0];
-
-    // Link each agg to the next agg. This will be
-    // to identify the next bucket aggregation
-    buckets = arrayToLinkedList(buckets);
-
-    // Create the raw data to be used in the spy panel
-    const raw = createRawData(vis, resp);
+  return function (table) {
+    let slices;
+    const names = {};
+    if (table.columns.length === 1) {
+      slices = [{ name: table.columns[0].title, size: table.rows[0][0].value }];
+      names[table.columns[0].title] = table.columns[0].title;
+    } else {
+      let parent;
+      slices = [];
+      table.rows.forEach(row => {
+        let dataLevel = slices;
+        // we always have one bucket column and one metric column (for every level)
+        for (let columnIndex = 0; columnIndex < table.columns.length; columnIndex += 2) {
```

Comment:
Do we need a check here to be sure that here `table.columns` contains always an even number of columns? 
What do you think? is somewhat assured only by the editor configuration right?

Reply:
its assured by visualisation type. Pie chart is the only one using hierarchical response handler. It defines isHierarchical to true, which will cause the courier to construct a request with metrics on every level.